tests/ipsecondary_test.py

```diff
@@ -0,0 +1,537 @@
+#!/usr/bin/env python3
+
+import tempfile
+import subprocess
+import json
+import multiprocessing
+import contextlib
+import logging
+import argparse
+
+from os import path
+from os import urandom
+from os import devnull
+from os import getcwd, chdir
+from uuid import uuid4
+from functools import wraps
+
+from fake_http_server.fake_test_server import ReUseHTTPServer as BaseTestServer
+
+logger = logging.getLogger("IPSecondaryTest")
+
+
+# The following are candidates for fixtures/common/shared components that can be reused by
+# multiple tests/test suits
+class UptaneTestRepo:
+
+    @property
+    def root_dir(self):
+        return self._repo_root_dir.name
+
+    @property
+    def image_rel_dir(self):
+        return 'repo/image'
+
+    @property
+    def target_rel_dir(self):
+        return 'repo/image/targets'
+
+    @property
+    def image_dir(self):
+        return path.join(self.root_dir, self.image_rel_dir)
+
+    @property
+    def target_dir(self):
+        return path.join(self.root_dir, self.target_rel_dir)
+
+    @property
+    def target_file(self):
+        return path.join(self.image_dir, 'targets.json')
+
+    def add_image(self, id, image_filename, target_name=None, image_size=1024):
+
+        targetname = target_name if target_name else image_filename
+
+        with open(path.join(self.image_dir, image_filename), 'wb') as image_file:
+            image_file.write(urandom(image_size))
+
+        subprocess.run([self._repo_manager_exe, '--path', self.root_dir,
+                        '--command', 'image', '--filename', image_filename, '--targetname', targetname],
+                       cwd=self.image_dir, check=True)
+
+        # update the director metadata
+        subprocess.run([self._repo_manager_exe, '--path', self.root_dir,
+                        '--command', 'addtarget', '--targetname', targetname,
+                        '--hwid', id[0], '--serial', id[1]], check=True)
+
+        # sign so the image becomes available for an update for a client/device
+        subprocess.run([self._repo_manager_exe, '--path', self.root_dir, '--command', 'signtargets'], check=True)
+
+        with open(self.target_file, "r") as target_file:
+            targets = json.load(target_file)
+            target_hash = targets["signed"]["targets"][targetname]["hashes"]["sha256"]
+
+        return target_hash
+
+    def __init__(self, repo_manager_exe):
+        self._repo_manager_exe = repo_manager_exe
+        self._repo_root_dir = tempfile.TemporaryDirectory()
+
+    def __enter__(self):
+        self._repo_root_dir.__enter__()
+        self._generate_repo()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._repo_root_dir.__exit__(exc_type, exc_val, exc_tb)
+
+    def _generate_repo(self):
+        subprocess.run([self._repo_manager_exe, '--path', self.root_dir, '--command', 'generate'], check=True)
+
+
+class UptaneTestBackend:
+
+    @property
```

Comment:
style question: what's the benefit of using this `@property` indirection here, compared to just defining `port` and `repo` members? Same question for constants like `image_rel_dir` of `UptaneTestRepo`.

Also, I've seen most __init__ of python classes declared at the top, is it also deliberate to have them in the middle of other methods?

Reply:
> what's the benefit of using this @property indirection here, compared to just defining port and repo members?

make public API explicitly defined, so users/clients don't need to learn implementation in order to figure our a class interface/API

Reply:
> Also, I've seen most init of python classes declared at the top, is it also deliberate to have them in the middle of other methods?

Yes, your're right, the standard approach/style is to put __init__ at the beginning of a class definition, I'll update my code correspondingly.

Reply:
Just updated the PR, __init__ was moved up and useless properties are  removed

Reply:
I don't want to insist on doing it at all cost but I also fail to see the changes right now. Github only shows a rebase without code changes when inspecting the last force-push.

Reply:
Your comment is reasonable.
The code should be updated now, somehow I screwed up to commit the changes in the first place.

Reply:
Ok now I see it :+1: